scanner/scanner.go

```diff
@@ -284,49 +321,27 @@ func (s *Scanner) Scan(ctx context.Context, foundCert func(*ct.LogEntry), foundP
 	s.unparsableEntries = 0
 	s.entriesWithNonFatalErrors = 0
 
-	latestSth, err := s.logClient.GetSTH(ctx)
+	sth, err := s.logClient.GetSTH(ctx)
 	if err != nil {
 		return fmt.Errorf("failed to GetSTH(): %v", err)
 	}
-	s.Log(fmt.Sprintf("Got STH with %d certs", latestSth.TreeSize))
+	s.Log(fmt.Sprintf("Got STH with %d certs", sth.TreeSize))
 
-	// TODO: cleanup Ticker and goroutine on return.
-	ticker := time.NewTicker(time.Second)
 	startTime := time.Now()
+	stop := make(chan bool)
+	go s.logThroughput(int64(sth.TreeSize), stop)
+	defer func() {
+		stop <- true
+		close(stop)
+	}()
+
 	fetches := make(chan fetchRange)
 	jobs := make(chan entryInfo, s.opts.BufferSize)
-	go func() {
-		slidingWindow := make([]int64, 15)
-		i, previousCount := 0, int64(0)
-		for range ticker.C {
-			certsMatched := atomic.LoadInt64(&s.certsMatched)
-			certsProcessed := atomic.LoadInt64(&s.certsProcessed)
-			slidingWindow[i%15], previousCount = certsProcessed-previousCount, certsProcessed
-
-			windowTotal := int64(0)
-			for _, v := range slidingWindow {
-				windowTotal += v
-			}
-			windowSeconds := 15
-			if i < 15 {
-				windowSeconds = i + 1
-			}
-
-			throughput := float64(windowTotal) / float64(windowSeconds)
-			remainingCerts := int64(latestSth.TreeSize) - int64(s.opts.StartIndex) - certsProcessed
-			remainingSeconds := int(float64(remainingCerts) / throughput)
-			remainingString := humanTime(remainingSeconds)
-			s.Log(fmt.Sprintf("Processed: %d certs (to index %d), matched %d (%2.2f%%). Throughput (last 15s): %3.2f ETA: %s\n",
-				certsProcessed, s.opts.StartIndex+int64(certsProcessed),
-				certsMatched, (100.0*float64(certsMatched))/float64(certsProcessed),
-				throughput, remainingString))
-			i++
-		}
-	}()
 
 	var ranges list.List
-	for start := s.opts.StartIndex; start < int64(latestSth.TreeSize); {
-		end := min(start+int64(s.opts.BatchSize), int64(latestSth.TreeSize)) - 1
+	// TODO(pavelkalinnikov): Add EndIndex parameter.
```

Comment:
See also: https://github.com/google/certificate-transparency-go/pull/172/commits/60abe5138a06f930659db8c8ad353c6027da18fd
(part of something else that's been pending for a while)

Reply:
I think these two can be complementary. If `EndIndex == -1` (by default), scan as usual up to `sth.TreeSize`, otherwise scan up to `min(sth.TreeSize, EndIndex)`. In both cases your `ScanLog` method will return the end of the scanned interval.

Reply:
See #199